Add ROS 2 CLI module to Fast DDS Discovery Server Tutorial (#3327)
ros2/ros2_documentation: source/Tutorials/Advanced/Discovery-Server/Discovery-Server.rst

ROS 2 Introspection
-------------------

> `ROS 2 Command Line Interface <https://github.com/ros2/ros2cli>`__ supports several introspection tools to analyze the behavior of a ROS 2 network.

Reviewer: ```suggestion
The `ROS 2 Command Line Interface <https://github.com/ros2/ros2cli>`__ supports several introspection tools to analyze the behavior of a ROS 2 network.
```

These tools (i.e. ``ros2 bag record``, ``ros2 topic list``, etc.) are very helpful to understand a ROS 2 working network.

Most of these tools use DDS simple discovery to exchange topic information with every existing participant (using simple discovery, every participant in the network is connected with each other).
However, the new Discovery Server v2 implements a network traffic reduction scheme that limits the discovery data between participants that do not share a topic.

> This means that not every node will receive every topic's discovery data unless it has a reader for that topic.

Reviewer: ```suggestion
This means that nodes will only a topic's discovery data if it has a reader for that topic.
```

Author: I made a little reword here. Hope you agree with it :smile:

> As most ROS 2 CLIs need a node in the network (some of them rely on a running ROS 2 daemon, and some create their own nodes), using the Discovery Server v2 these nodes will not have all the network information, and thus their functionalities are limited.

Reviewer: ```suggestion
As most ROS 2 CLIs need a node in the network (some of them rely on a running ROS 2 daemon, and some create their own nodes), using the Discovery Server v2 these nodes will not have all the network information, and thus their functionality will be limited.
```

> The Discovery Server v2 functionality allows every Participant running as a **SUPER_CLIENT**, a kind of **Client** that connects to a **SERVER**, from which it receives all the available discovery information (instead of just what it needs).

Reviewer: ```suggestion
The Discovery Server v2 functionality allows every Participant to run as a **Super Client**, a kind of **Client** that connects to a **Server**, from which it receives all the available discovery information (instead of just what it needs).
```

In this sense, ROS 2 introspection tools can be configured as **Super Client**, thus being able to discover every entity that is using the Discovery Server protocol within the network.

.. note::

    In this section we use the term *Participant* as a DDS entity. Each DDS *Participant* corresponds with a ROS 2 *Context*, a ROS 2 abstraction over DDS.
    `Nodes <ROS2Nodes>` are ROS 2 entities that rely on DDS communication interfaces: ``DataWriter`` and ``DataReader``.
    Each *Participant* can hold multiple ROS 2 Nodes.
    For further details about these concepts, please visit the `Node to Participant mapping design document <http://design.ros2.org/articles/Node_to_Participant_mapping.html>`__

Daemon's related tools
^^^^^^^^^^^^^^^^^^^^^^

The ROS 2 Daemon is used in several ROS 2 CLI introspection tools.
It creates its own Participant to add a ROS 2 Node to the network graph, in order to receive all the data sent.
In order for the ROS 2 CLI to work when using Discovery Server mechanism, the ROS 2 Daemon needs to be
configured as **Super Client**.
Therefore, this section is devoted to explain how to use ROS 2 CLI with ROS 2 Daemon running as a **Super Client**.

> This will allow the Daemon to discover the entire Node graph, and to receive every topic and endpoint information.

Reviewer: ```suggestion
This will allow the Daemon to discover the entire Node graph, and to receive all topic and endpoint information.
```

ROS 2 Introspection
-------------------

> ROS 2 Command Line Interface (CLI) implements several introspection tools to analyze the behaviour of a ROS 2 network.

Reviewer 2: ```suggestion
`ROS 2 Command Line Interface <https://github.com/ros2/ros2cli>`__ supports several introspection tools to analyze the behavior of a ROS 2 network.
```

Author: Applied

These tools (i.e. ``ros2 bag record``, ``ros2 topic list``, etc.) are very helpful to understand a ROS 2 working network.

Most of these tools use DDS simple discovery to exchange topic information with every existing participant (using simple discovery, every participant in the network is connected with each other).
However, the new Discovery Server v2 implements a network traffic reduction scheme that limits the discovery data between participants that do not share a topic.

> This means that not every node will receive every topic's discovery data unless it has a reader for that topic.

Reviewer 2: Off topic from this PR, but this is also supported as general ROS 2 or Fast-DDS participant feature? i think this feature is like `Lazy Discovery` to reduce the discovery traffic unless it is necessary?

Author: Yes it is supported.
This avoids all the discovery traffic that is not needed in the application. As two participants that do not interact doesn't need to know each other. But discovery server knows everyone. You can read more about it here: https://fast-dds.docs.eprosima.com/en/v2.9.1/fastdds/discovery/discovery_server.html and here: https://eprosima-discovery-server.readthedocs.io/en/latest/user_manual/getting_started/getting_started.html.

    `Nodes <ROS2Nodes>` are ROS 2 entities that rely on DDS communication interfaces: ``DataWriter`` and ``DataReader``.
    Each *Participant* can hold multiple ROS 2 Nodes.
    For further details about these concepts, please visit the `Node to Participant mapping design document <http://design.ros2.org/articles/Node_to_Participant_mapping.html>`__

Daemon's related tools
^^^^^^^^^^^^^^^^^^^^^^

The ROS 2 Daemon is used in several ROS 2 CLI introspection tools.
It creates its own Participant to add a ROS 2 Node to the network graph, in order to receive all the data sent.
In order for the ROS 2 CLI to work when using Discovery Server mechanism, the ROS 2 Daemon needs to be
configured as **Super Client**.
Therefore, this section is devoted to explain how to use ROS 2 CLI with ROS 2 Daemon running as a **Super Client**.
This will allow the Daemon to discover the entire Node graph, and to receive every topic and endpoint information.
To do so, a Fast DDS XML configuration file is used to configure the ROS 2 Daemon and CLI tools.

.. warning::
    Although it is possible to run the ROS 2 Daemon as a **Server**, this is not recommended since the daemon will stop

>     after two hours of inactivity, taking the **Server** down with it.

Reviewer 2: with current implementation, `ros2 daemon start` ROS 2 daemon will NOT stop running even after 2 hours. (that used to be but AFAIK it runs forever now until explicitly stopped by user application.)

see https://github.com/ros2/ros2cli/pull/652, especially here https://github.com/ros2/ros2cli/pull/652/files#r669866793

Author: Nice. That warning has been erased.
